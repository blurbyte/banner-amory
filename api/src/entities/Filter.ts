import { Item, Tier } from './Item';

// 0 means ALL parts and ranks
export type Filter = {
  part: 0 | Item['gamePart'];
  rank: 0 | Item['rank'];
  bonus: FilterBonus;
  tier: FilterTier;
};

export enum FilterBonus {
  Any = 'ANY',
  MainStat = 'MAIN_STAT',
  Talent = 'TALENT',
  PerTurn = 'PER_TURN',
  OnRest = 'ON_REST',
  Chance = 'CHANCE',
  FromMarket = 'FROM_MARKET',
  Movement = 'MOVEMENT'
}

export type FilterTier = 'ANY' | Tier;

// Item properties on which filtering is based
export type AvailableFilterBonusProperty = Extract<
  keyof Item,
  | 'statStrength'
  | 'statWillpower'
  | 'statArmor'
  | 'statArmorBreak'
  | 'talentAllRanks'
  | 'talentStrength'
  | 'talentWillpower'
  | 'talentArmor'
  | 'talentArmorBreak'
  | 'talentExertion'
  | 'perTurnArmor'
  | 'perTurnWillpower'
  | 'onRestArmor'
  | 'onRestWillpower'
  | 'chanceCrit'
  | 'chanceDodge'
  | 'chanceDivert'
  | 'acquirementMarketplace'
  | 'movement'
>;
