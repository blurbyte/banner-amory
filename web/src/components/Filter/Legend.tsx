import styled from 'styled-components';
import BaseLabel from '../Label';

const Paragraph = BaseLabel.withComponent('p');

const Legend = styled(Paragraph)`
  margin: 0;
  margin-right: 1rem;
  line-height: 2.4rem;
`;

export default Legend;
