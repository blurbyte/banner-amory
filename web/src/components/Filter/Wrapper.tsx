import styled from 'styled-components';

const Wrapper = styled.div`
  display: flex;
  margin-right: 2rem;
  margin-top: 1rem;
`;

export default Wrapper;
