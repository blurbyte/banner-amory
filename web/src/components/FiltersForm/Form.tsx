import styled from 'styled-components';

type FormProps = {
  isExpanded?: boolean;
};

const Form = styled.form<FormProps>`
  overflow: hidden;
  height: ${props => (props.isExpanded ? '14rem' : 0)};
  transition: height 0.25s ease-in-out;
  padding-left: 10rem;
`;

export default Form;
