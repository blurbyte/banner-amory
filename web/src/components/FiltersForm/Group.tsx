// simple wrapper to group filters

import styled from 'styled-components';

const Group = styled.div`
  display: flex;
  flex-wrap: wrap;

  &:not(:first-child) {
    margin-top: 1.2rem;
  }
`;

export default Group;
