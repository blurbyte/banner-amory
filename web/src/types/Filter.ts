// for sake of simplicity input values are strings and compability with GraphQL

export type Filter = {
  part: 'all' | '1' | '2';
  rank: 'all' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10';
  bonus: FilterBonus;
};

export enum FilterBonus {
  Any = 'ANY',
  MainStat = 'MAIN_STAT',
  Talent = 'TALENT',
  PerTurn = 'PEWR_TURN',
  OnRest = 'ON_REST',
  Chance = 'CHANCE',
  FromMarket = 'FROM_MARKET',
  Movement = 'MOVEMENT'
}
