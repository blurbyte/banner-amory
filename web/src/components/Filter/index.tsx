// group of filter items
// with automatic name, onChange and checkedValue applied to each child

import * as React from 'react';

import { Filter as FilterType } from '@sharedTypes/Filter';
import FilterItem from '../FilterItem';
import Wrapper from './Wrapper';
import Legend from './Legend';

export type FilterProps = {
  children: React.ReactNode;
  name: keyof FilterType;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  checkedValues: FilterType;
  label?: string;
};

class Filter extends React.Component<FilterProps> {
  static Item = FilterItem;

  renderChildren = () => {
    const { children, name, onChange, checkedValues } = this.props;
    let checkedValue = checkedValues[name];

    // pass name prop to all Item children
    return React.Children.map(children, child => {
      return React.cloneElement(child as React.ReactElement<any>, { name, onChange, checkedValue });
    });
  };

  render() {
    const { name, label } = this.props;

    return (
      <Wrapper>
        <Legend>{label ? label : name}</Legend>
        {this.renderChildren()}
      </Wrapper>
    );
  }
}

export default Filter;
