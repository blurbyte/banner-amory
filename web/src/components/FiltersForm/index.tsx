import * as React from 'react';

import { Filter as FilterType, FilterBonus } from '@sharedTypes/Filter';
import Filter from '../Filter';
import Form from './Form';

type FiltersFormProps = {
  isExpanded?: boolean;
};

type FiltersFormState = Readonly<FilterType>;

class FiltersForm extends React.Component<FiltersFormProps, FiltersFormState> {
  readonly state: FiltersFormState = {
    part: 'all',
    rank: 'all',
    bonus: FilterBonus.Any
  };

  render() {
    const { isExpanded } = this.props;

    return (
      <Form isExpanded={isExpanded}>
        <Filter name="part" onChange={this.handleChange} checkedValues={this.state}>
          <Filter.Item value="all" />
          <Filter.Item value="1" />
          <Filter.Item value="2" />
        </Filter>
        <Filter name="rank" onChange={this.handleChange} checkedValues={this.state}>
          <Filter.Item value="all" />
          <Filter.Item value="1" />
          <Filter.Item value="2" />
          <Filter.Item value="3" />
          <Filter.Item value="4" />
          <Filter.Item value="5" />
          <Filter.Item value="6" />
          <Filter.Item value="7" />
          <Filter.Item value="8" />
          <Filter.Item value="9" />
          <Filter.Item value="10" />
        </Filter>
        <Filter name="bonus" onChange={this.handleChange} checkedValues={this.state}>
          <Filter.Item value={FilterBonus.Any} label="any" />
          <Filter.Item value={FilterBonus.MainStat} label="main stat" />
          <Filter.Item value={FilterBonus.Talent} label="talent" />
          <Filter.Item value={FilterBonus.PerTurn} label="per turn" />
          <Filter.Item value={FilterBonus.OnRest} label="on rest" />
          <Filter.Item value={FilterBonus.Chance} label="chance" />
          <Filter.Item value={FilterBonus.FromMarket} label="from market" />
          <Filter.Item value={FilterBonus.Movement} label="movement" />
        </Filter>
      </Form>
    );
  }

  private handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;

    this.setState({
      [name]: value
    } as FiltersFormState);
  };
}

export default FiltersForm;
