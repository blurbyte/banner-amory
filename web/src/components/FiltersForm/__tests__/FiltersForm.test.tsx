import * as React from 'react';
import * as renderer from 'react-test-renderer';
import 'jest-styled-components';

import { FilterBonus, OrderBy } from '@sharedTypes/Filter';
import FiltersForm from '../index';

test('renders correctly', () => {
  const component = renderer.create(
    <FiltersForm
      changeFilter={jest.fn()}
      checkedValues={{ part: 1, rank: 1, bonus: FilterBonus.Any, orderBy: OrderBy.Rank }}
    />
  );
  expect(component.toJSON()).toMatchSnapshot();
});
