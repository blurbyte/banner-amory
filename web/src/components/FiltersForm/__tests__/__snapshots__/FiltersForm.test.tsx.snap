// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 2.4rem;
  line-height: 2.4rem;
  min-width: 2.4rem;
  background: rgba(61,38,218,0.2);
  color: #ffffff;
  font-weight: 400;
  cursor: pointer;
  padding: 0 0.6rem;
  margin-right: 1px;
  text-transform: capitalize;
  -webkit-transition: background 0.15s linear;
  transition: background 0.15s linear;
}

.c7 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 2.4rem;
  line-height: 2.4rem;
  min-width: 2.4rem;
  background: #d92f29;
  color: #ffffff;
  font-weight: 400;
  cursor: pointer;
  padding: 0 0.6rem;
  margin-right: 1px;
  text-transform: capitalize;
  -webkit-transition: background 0.15s linear;
  transition: background 0.15s linear;
}

.c6 {
  position: absolute;
  top: 0;
  left: 0;
  visibility: hidden;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 2rem;
  margin-top: 1rem;
}

.c4 {
  font-size: 1.2rem;
  color: rgba(61,38,218,0.5);
  text-transform: uppercase;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
}

.c3 {
  margin: 0;
  margin-right: 1rem;
  line-height: 2.4rem;
}

.c0 {
  overflow: hidden;
  height: 0;
  -webkit-transition: height 0.25s ease-in-out;
  transition: height 0.25s ease-in-out;
  padding-left: 10rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c1:not(:first-child) {
  margin-top: 1.2rem;
}

<form
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <p
        className="c3 c4"
      >
        part
      </p>
      <label
        checked={false}
        className="c5"
      >
        all
        <input
          checked={false}
          className="c6"
          name="part"
          onChange={[Function]}
          type="radio"
          value={0}
        />
      </label>
      <label
        checked={true}
        className="c7"
      >
        1
        <input
          checked={true}
          className="c6"
          name="part"
          onChange={[Function]}
          type="radio"
          value={1}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        2
        <input
          checked={false}
          className="c6"
          name="part"
          onChange={[Function]}
          type="radio"
          value={2}
        />
      </label>
    </div>
    <div
      className="c2"
    >
      <p
        className="c3 c4"
      >
        rank
      </p>
      <label
        checked={false}
        className="c5"
      >
        all
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={0}
        />
      </label>
      <label
        checked={true}
        className="c7"
      >
        1
        <input
          checked={true}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={1}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        2
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={2}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        3
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={3}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        4
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={4}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        5
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={5}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        6
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={6}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        7
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={7}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        8
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={8}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        9
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={9}
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        10
        <input
          checked={false}
          className="c6"
          name="rank"
          onChange={[Function]}
          type="radio"
          value={10}
        />
      </label>
    </div>
    <div
      className="c2"
    >
      <p
        className="c3 c4"
      >
        bonus
      </p>
      <label
        checked={true}
        className="c7"
      >
        any
        <input
          checked={true}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="ANY"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        main stat
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="MAIN_STAT"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        talent
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="TALENT"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        per turn
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="PER_TURN"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        on rest
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="ON_REST"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        chance
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="CHANCE"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        from market
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="FROM_MARKET"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        movement
        <input
          checked={false}
          className="c6"
          name="bonus"
          onChange={[Function]}
          type="radio"
          value="MOVEMENT"
        />
      </label>
    </div>
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <p
        className="c3 c4"
      >
        order by
      </p>
      <label
        checked={true}
        className="c7"
      >
        rank
        <input
          checked={true}
          className="c6"
          name="orderBy"
          onChange={[Function]}
          type="radio"
          value="RANK"
        />
      </label>
      <label
        checked={false}
        className="c5"
      >
        alphabetically
        <input
          checked={false}
          className="c6"
          name="orderBy"
          onChange={[Function]}
          type="radio"
          value="ALPHABETICALLY"
        />
      </label>
    </div>
  </div>
</form>
`;
